(WIP) Workflows for common comparisons and misc additions

mranst opened the pull request:
## Description

This PR lays the groundwork for introducing workflows for common comparisons, and also introduces features such as email messaging and the option to set certain tasks for pause #490 .

Two rudimentary suites based on discussions in #442 have been introduced. `compare_variational` runs a simple fgrep command to find the residual norms in jedi log output. `compare_jedi` runs ctests on jedi builds and runs a crude comparison of which ones passed. To make this work, I introduced two additional options to the `swell task` command, an `iteration` flag which can be used to differentiate the same tasks being run on different experiments, and an `output` option which allows tasks that use it to send their test output to different locations (such as the comparison experiment, rather than the individual swell experiments that are being compared). These comparison suites make use of the updated cylc templating using python.

A command has been added `swell compare <type> (experiment locations)` to easily create comparison suites

Users can now select certain tasks to pause the workflow after by specifying a list in the overrides. This should allow for adjustments when developing/debugging. Using cylc's native capability for jinja2 templating, this can be toggled on when running `swell launch` using the flag `-d`.

Options have been added for sending email messages using cylc. This should theoretically work, according to cylc's syntax and example's I've seen from Doruk and Seth, but for some reason I don't actually receive email messages. Maybe someone else can try it. 

Adding a file `~/.swell/swell-settings.yaml` with the entry `email_address = <email address>`, and setting the configuration option `cylc_message_parameters` will add the language to flow.cylc. Email message won't be set unless they are toggled on using the flag `-m`, for convenience. All tasks will have message on failure/submit-failure preference by default, or can be set with custom events using a dictionary.

mranst committed:
Initial commit

mranst committed:
adding workflow string

mranst committed:
working

mranst committed:
fixing section formatting

mranst committed:
redo task questions

mranst committed:
update task questions

mranst committed:
update task questions again

mranst committed:
mostly working

mranst committed:
Functional, fixing indentation

mranst committed:
pycodestyle fixes

mranst committed:
Fix slurm formatting in runtime

mranst committed:
Simplify task runtime definitions

mranst committed:
remove print statement

mranst committed:
Fix slurm setting

mranst committed:
Remove duplicate import

mranst committed:
Remove setter methods

mranst committed:
Simplify formatting

mranst committed:
Simplify workflows and suite configs

mranst committed:
Clean up and introduce question order test

mranst committed:
Initial commit

mranst committed:
Merge branch 'se/mranst/cylc_templating' into feature/mranst/workflows_for_comparison

mranst committed:
remove non-task tasks from tasks

mranst committed:
Merge branch 'se/mranst/cylc_templating' into feature/mranst/workflows_for_comparison

mranst committed:
add test for jedi

mranst committed:
add compare command for jedi

mranst committed:
Add workflow pause and event status emails

mranst committed:
Add message on pause and fix to messaging

mranst committed:
pycodestyle fix

mranst committed:
Update slurm code test

mranst committed:
added some suites

mranst committed:
add workflows

mranst committed:
add all suite workflows

mranst committed:
fix cycle times

mranst committed:
add all suites

mranst committed:
fixes

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst committed:
add cycling_varbc

mranst committed:
remove print statement

mranst committed:
revise templating

mranst committed:
revise templating for all suites

mranst committed:
fixes for code tests

mranst committed:
try to fix unused variables

mranst committed:
functional code tests

mranst committed:
pycodestyle fixes

mranst committed:
Merge branch 'tmp' into feature/mranst/workflows_for_comparison

mranst committed:
revise templating for 3dfgat_cycle

mranst committed:
fix task_runtimes

mranst committed:
fix cycling

mranst committed:
fix for forecast_geos

mranst committed:
fixes to suites

mranst committed:
add comments

mranst committed:
test plots

mranst committed:
add imcrement comparison

mranst committed:
Functional comparison plots

mranst committed:
Add comments for comparison plots

mranst committed:
pycodestyle fixes

mranst commented:
Every suite has been incorporated into this new method, and the code tests have been updated to work. @Dooruk I've added a task that runs eva comparison plots for soca increments, if we the way this works I'll work on adding more. If you have any suggestions for the formatting let me know, I can't figure out how to get the title centered, for example

<img width="437" height="934" alt="inc_Temp" src="https://github.com/user-attachments/assets/d295fa3e-d38f-43e3-b404-06838b36a430" />

Dooruk commented:
Can I test this directly from `feature/mranst/workflows_for_comparison`?

mranst commented:
> Can I test this directly from `feature/mranst/workflows_for_comparison`?

Yes, you can do `swell create compare_3dvar`, and manually enter locations for `experiment.yaml` for two experiments

Dooruk commented:
Thanks @mranst, substantial work! I'm looking into this now, I couldn't get it to run but a few questions:

- Is the procedure for comparisons as follows: 
1)`swell compare jedi | variational`, create experiment folder
2) modify `experiment.yaml`, add two experiment paths
3) swell launch
?

- This also includes flow.cylc changes right? Should I try to understand those changes from the other PR first?
- Two issues i see might be; 1) how to handle if two experiments were run on different times, years say. We could either explicitly define the time period for comparisons, or expect Swell to figure that out automatically by parsing `run` directories. 2) This is definitely less obvious but two experiments might have different data assimilation windows (this is only the case for marine workflows, they could be 6-hr,24-hr or 5 days). I think for simplicity we should warn the user and fail the suite if there is mismatch.

mranst commented:
> Thanks @mranst, substantial work! I'm looking into this now, I couldn't get it to run but a few questions:
> 
>     * Is the procedure for comparisons as follows:
>       1)`swell compare jedi | variational`, create experiment folder
> 
> 
>     2. modify `experiment.yaml`, add two experiment paths
> 
>     3. swell launch
>        ?
> 
> 
>     * This also includes flow.cylc changes right? Should I try to understand those changes from the other PR first?
> 
>     * Two issues i see might be; 1) how to handle if two experiments were run on different times, years say. We could either explicitly define the time period for comparisons, or expect Swell to figure that out automatically by parsing `run` directories. 2) This is definitely less obvious but two experiments might have different data assimilation windows (this is only the case for marine workflows, they could be 6-hr,24-hr or 5 days). I think for simplicity we should warn the user and fail the suite if there is mismatch.

I would specify the two file paths in an override file, run `swell create compare_3dvar -o override.yaml`. You can also add the filepaths after the fact and do `swell task EvaComparisonIncrement -m geos_marine experiment.yaml`.

Yeah handling the cycle times between two experiments is one of the more challenging aspects to this. I kept it very simple in this task, just parsing the common dates in the run directory, but the logic around matching windows should definitely be more fleshed out

The flow.cylc templating and all this other stuff kind of got wrapped up into one big PR, but this eva task doesn't necessarily need to be included in that since that logic is all handled within the task

Dooruk commented:
> 
> The flow.cylc templating and all this other stuff kind of got wrapped up into one big PR, but this eva task doesn't necessarily need to be included in that since that logic is all handled within the task

Ok, in that sense should I review https://github.com/GEOS-ESM/swell/pull/540 first or this one?

mranst commented:
> > The flow.cylc templating and all this other stuff kind of got wrapped up into one big PR, but this eva task doesn't necessarily need to be included in that since that logic is all handled within the task
> 
> Ok, in that sense should I review #540 first or this one?

That one is far outdated, I closed it

Dooruk commented:
Ok, I tried providing two experiment paths, suite immediately stops after `swell launch`. i can take a look tomorrow:

```yaml
# Datetime this file was created (auto added)
datetime_created: 20250717_165317Z

# Computing platform to run the experiment
platform: nccs_discover_sles15

# Record of the suite being executed
suite_to_run: compare_variational

# What is the experiment id?
experiment_id: swell-compare_variational

# What is the experiment root (the directory where the experiment will be stored)?
experiment_root: /discover/nobackup/dardag/SwellExperiments

# Provide paths to two experiments to run comparison tests on.
comparison_experiment_paths: ['/discover/nobackup/projects/gmao/advda/dardag/sles15/fgat_odasemc_01','/discover/nobackup/projects/gmao/advda/dardag/sles15/fgat_jra55_01']

# Enter the model components for this model.
model_components:
- geos_marine

# Since this suite is non-cycling choose how many hours the workflow can run ahead?
runahead_limit: P4

# Provide a dictionary mapping tasks to cylc event statuses, or 'auto' to automatically configure these based on the graph.
task_email_parameters: auto

# Specify any tasks that the workflow should pause on (for development purposes).
pause_on_tasks: []

# Configurations for the model components.
models:

  # Configuration for the geos_marine model component.
  geos_marine:

    # What is the duration between the middle of the window and the beginning?
    window_offset: PT6H

    # Do you want to use a 3D or 4D (including FGAT) window?
    window_type: 3D
```

mranst committed:
fix workflow

mranst commented:
There was a bug, it should be good now

Dooruk commented:
So I'm trying to test this, I think I made some progress. It's a bit hard to follow with the flow.cylc changes.

Is compare 3dvar meant to obtain some flow.cylc configuration from the `3dvar` suite?

`compare_3dvar` only picked the last 6 cycles from overlapping cycle times for the following two experiments, and didn't grep the outputs..

<img width="1371" height="268" alt="Screenshot 2025-07-18 at 3 32 31 PM" src="https://github.com/user-attachments/assets/02e238d9-f2ca-4ce8-8c8f-59108da8c4fe" />

mranst committed:
check for fgrep status

mranst commented:
> So I'm trying to test this, I think I made some progress. It's a bit hard to follow with the flow.cylc changes.
> 
> Is compare 3dvar meant to obtain some flow.cylc configuration from the `3dvar` suite?
> 
> `compare_3dvar` only picked the last 6 cycles from overlapping cycle times for the following two experiments, and didn't grep the outputs..
> <img alt="Screenshot 2025-07-18 at 3 32 31 PM" width="1371" height="268" src="https://private-user-images.githubusercontent.com/38666458/468187598-02e238d9-f2ca-4ce8-8c8f-59108da8c4fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI4NzAzNTAsIm5iZiI6MTc1Mjg3MDA1MCwicGF0aCI6Ii8zODY2NjQ1OC80NjgxODc1OTgtMDJlMjM4ZDktZjJjYS00Y2U4LThjOGYtNTkxMDhkYThjNGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE4VDIwMjA1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxZjkwMmRkYWIzZjI3MzlhMmFhNzU4YTkxY2ZmMzc5NGJlMmVhNGM5ZGE2YzQxNWQ4N2UwZjAyMmE3MjJjNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bDUsgPBKoV1X__IsP3tMsrjbvR-jqgxoKNVR3e_xmHc">

compare_3dvar takes information about window types and such from the suite config for 3dvar, that way it can be used in the task (like in the regular EvaIncrement task which runs differently for different suite parameters). But right now, the flow.cylc templating for the comparison suite is its own thing. 

It sounds like there's an error when it tries to grep the values, maybe it's trying to run on one cycle but not the other. I've added a check so it should now tell you what the problem is

Right now it'll only run the comparison plots for overlapping cycles, are there ones that it's missing from those?

Dooruk commented:
Ok, I'm having a better grasp of this so I will suggest a few key changes. This is surely a complicated development, will take a few iterations to make it more flexible/expendable, and is not trivial so let me know if you need hands on help or more clarification.

- I think finding the common timesteps is the essential part of all of these workflow comparisons business. So I suggest taking common timesteps function out of EvaCompare task, make it a utility to be used in the suites level, maybe under `all_suites.py`. Because there will be different types of suites going forward. Then, it can be a two step feature.

  Step 1: User defines start/finish cycle times
  Step 2: Swell figures out overlapping cycles,  between user time bounds and if none exists warns the user.

  While I was testing your PR, Swell struggled finding the common timesteps. I see you look at the timesteps under the `run` directory which is the right approach but while doing comparisons you might want to use datetime objects so you can order them properly.
  
- Maybe not in this PR, but going forward we may want to copy/link saved experiment outputs (whether from R2D2 or wherever @rtodling stores his experiments) to comparisons cycle directories. With my cycling tests, I save restarts or backgrounds and erase certain files, such as the increments. With R2D2 it would be easy, since it is standardized but without it that's more tricky. He did provide some bits of information here https://github.com/GEOS-ESM/swell/issues/563

  Below is what I do with `eva_timeseries` suite:
  https://github.com/GEOS-ESM/swell/pull/549

- `FGrepResidualNorm` is one type of an OOPS log parser but there could be more going forward since there is a lot more information inside the OOPS log. So I suggest giving this task a generic name like "oops_log_parser" (or someting more creative) and make this task consisting of different methods. I believe the proper SE term for this is "factory method" where you have an entry point function and have multiple subfunctions depending on what type of a function you would like utilize.

  This is what I'm doing in this following task and I can choose between `explicit_diffusion` or `bump` or add more options going forward.
  
  https://github.com/GEOS-ESM/swell/blob/839ff036a3270fe020cf046724ca0d016121c7e1/src/swell/tasks/generate_b_climatology.py#L50-L61
  
  The inputs could be implemented by a key in `experiment.yaml`, say `parser_options:`
  
-  There are some build in methods for what you are doing in a few places, e.g. `self.get_model()`, or `self.cycle_time_dto()`

metdyn committed:
Link akbk*.nc4 to the fv3files/ directory (#602)

rtodling committed:
SWELL now controls on peripheral files (#603)

mranst committed:
use cylc cycling

mranst committed:
rework fgrep task

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst commented:
@Dooruk I've made some updates to the cycle matching. You can choose a start and stop cycle point and cycling interval manually, or set them to None and it will automatically match overlapping intervals. It will then cycle like a normal swell suite. It'll check that the window length and offset are the same between the experiments. I also have a version of it that directly matches the directories like I had it before and runs the cycles via integer cycling, but I like this way better

mranst committed:
Add task to get observations from GEOS location (#577)

* Initial commit

* intermediate

* rename task and add toggle

* restore get_observations

* fix toggle

* Updated to link

* Update naming and add task to 3dvar_atmos

* add task

* Add task to other suites

* add runtime for getobsnotinr2d2

Dooruk commented:
A quick comment, the commented out part didn't work for me:

```py

    compare_3dvar = QuestionList(
        list_name="compare_3dvar",
        questions=[
            _3dvar,
            compare_variational,
            qd.model_components(['geos_marine']),
            qd.start_cycle_point(),
            # qd.start_cycle_point(default_value=None, widget_type=WidgetType.STRING),
            qd.final_cycle_point(),
            # qd.final_cycle_point(default_value=None, widget_type=WidgetType.STRING),
        ]
    )
```

I receive the following error:

```
    start_cycle_dto = max(start_cycle_dtos)
                      ^^^^^^^^^^^^^^^^^^^^^
ValueError: max() arg is an empty sequence

```

The parser task fails for me and I'm taking a closer look at that but the Cylc and the workflow aspect seems to be working.

mranst committed:
Fix for missing cycle times

mranst commented:
> A quick comment, the commented out part didn't work for me:
> 
> ```python
>     compare_3dvar = QuestionList(
>         list_name="compare_3dvar",
>         questions=[
>             _3dvar,
>             compare_variational,
>             qd.model_components(['geos_marine']),
>             qd.start_cycle_point(),
>             # qd.start_cycle_point(default_value=None, widget_type=WidgetType.STRING),
>             qd.final_cycle_point(),
>             # qd.final_cycle_point(default_value=None, widget_type=WidgetType.STRING),
>         ]
>     )
> ```
> 
> I receive the following error:
> 
> ```
>     start_cycle_dto = max(start_cycle_dtos)
>                       ^^^^^^^^^^^^^^^^^^^^^
> ValueError: max() arg is an empty sequence
> ```
> 
> The parser task fails for me and I'm taking a closer look at that but the Cylc and the workflow aspect seems to be working.

Do you have any experiments specified in override.yaml? This should only occur when it has nothing to search for, but I've pushed a fix for it

Dooruk reviewed (changes requested):
Ok I figured out the problem. In `JediOopsLogParser` `cycle_dir` is used, as such:

```
class JediOopsLogParser(taskBase):

    def fgrep_residual_norm(self):

        cycle_dir = self.cycle_dir()

```

this will not use the `run` folders from the original experiment folders but rather where you created your experiment directory.`cycle_dir` is defined in the base class.

Logs are located under at `comparison_exp1` and `comparison_exp2` `run`. For instance for my case i need my logs from the run/20210701T120000Z etc.

<img width="849" height="409" alt="Screenshot 2025-07-31 at 2 50 07 PM" src="https://github.com/user-attachments/assets/102a9796-feab-492d-88b0-7e5f35c2475e" />

There are many ways to create this but one way might be to utilize `comparison_experiment_paths` by dropping the last two levels and combining `run` and isofolder information:

`/discover/nobackup/projects/gmao/advda/dardag/sles15/fgat_jra55_01/fgat_jra55_01-suite/experiment.yaml`

One last suggestion for subprocess, `run_track_log_subprocess` via from swell.utilities.shell_commands import run_track_log_subprocess spits out the error so it made it easier for me to find out about the error. I see that you need to provide an additional kwarg  for `results = output.stdout` to work but it may still work?

OR 

other option is to link or copy log files to your current exp folder but I think that is unnecessary

mranst commented:
> Ok I figured out the problem. In `JediOopsLogParser` `cycle_dir` is used, as such:
> 
> ```
> class JediOopsLogParser(taskBase):
> 
>     def fgrep_residual_norm(self):
> 
>         cycle_dir = self.cycle_dir()
> ```
> 
> this will not use the `run` folders from the original experiment folders but rather where you created your experiment directory.`cycle_dir` is defined in the base class.
> 
> Logs are located under at `comparison_exp1` and `comparison_exp2` `run`. For instance for my case i need my logs from the run/20210701T120000Z etc.
> <img alt="Screenshot 2025-07-31 at 2 50 07 PM" width="849" height="409" src="https://private-user-images.githubusercontent.com/38666458/473137476-102a9796-feab-492d-88b0-7e5f35c2475e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM5ODk2NjgsIm5iZiI6MTc1Mzk4OTM2OCwicGF0aCI6Ii8zODY2NjQ1OC80NzMxMzc0NzYtMTAyYTk3OTYtZmVhYi00OTJkLTg4YjAtN2U1ZjM1YzI0NzVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzMxVDE5MTYwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0MGE2NjdiYTgxYWEzMGM1ZjlkMWM0M2Q4MzliNjcxMGE1MDE4MmYyMDNhZDhkNzFkNzdkNzExNTg0ZDA1NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JPdduUVQDBAXu4Xg8VyiHFskSIdZ93BF5Ipr6DJJKFk">
> 
> There are many ways to create this but one way might be to utilize `comparison_experiment_paths` by dropping the last two levels and combining `run` and isofolder information:
> 
> `/discover/nobackup/projects/gmao/advda/dardag/sles15/fgat_jra55_01/fgat_jra55_01-suite/experiment.yaml`
> 
> One last suggestion for subprocess, `run_track_log_subprocess` via from swell.utilities.shell_commands import run_track_log_subprocess spits out the error so it made it easier for me to find out about the error. I see that you need to provide an additional kwarg for `results = output.stdout` to work but it may still work?
> 
> OR
> 
> other option is to link or copy log files to your current exp folder but I think that is unnecessary

This might be overcomplicated, but I designed that task to essentially run as part of the existing experiments. That way, we could potentially have JediOopsLogParser run just as part as the 3dvar suite without necessarily being tied down to the comparison suite in particular. The workflow templating will direct that task to run on the specific experiment.yaml in the comparison experiment paths:

<img width="3010" height="726" alt="image" src="https://github.com/user-attachments/assets/f95b9597-0a69-426f-a17b-c807c4c6be71" />

Dooruk commented:
> This might be overcomplicated, but I designed that task to essentially run as part of the existing experiments. That way, we could potentially have JediOopsLogParser run just as part as the 3dvar suite without necessarily being tied down to the comparison suite in particular. The workflow templating will direct that task to run on the specific experiment.yaml in the comparison experiment paths
 
 Hmm I see, it does seem a bit complicated. So users have to specify that during the create process? Or would this task/suite automatically engage? 
 
I would think mostly folks would like to compare experiments after they run. For instance when we are testing Tier2 one test would be to compare runs with a static or stable JEDI build against the new one.

mranst commented:
> Hmm I see, it does seem a bit complicated. So users have to specify that during the create process? Or would this task/suite automatically engage?
> 
> I would think mostly folks would like to compare experiments after they run. For instance when we are testing Tier2 one test would be to compare runs with a static or stable JEDI build against the new one.

As long as the `comparison_experiment_paths` are specified in override.yaml, swell will handle setting up that task in flow.cylc automatically. Trying to alter the experiment paths after the creating the experiment and flow.cylc would be awkward. I can just alter the task to cycle through the paths by itself if that's too complicated

Dooruk commented:
> > Hmm I see, it does seem a bit complicated. So users have to specify that during the create process? Or would this task/suite automatically engage?
> > I would think mostly folks would like to compare experiments after they run. For instance when we are testing Tier2 one test would be to compare runs with a static or stable JEDI build against the new one.
> 
> As long as the `comparison_experiment_paths` are specified in override.yaml, swell will handle setting up that task in flow.cylc automatically. Trying to alter the experiment paths after the creating the experiment and flow.cylc would be awkward. I can just alter the task to cycle through the paths by itself if that's too complicated

Ok, maybe I'm not running this properly, let me read your description again.

mranst committed:
Automatically set window length for cycling (#604)

* Automatically set window length

* Update 3dfgat_cycle, remove comment

rtodling commented:
@mranst I have completely followed the discussions here - although I spent sometime this AM reading through them. 

I will have to say that I am rather surprised by the amount of changes this requires ... in my mind, I would think this should require no change to existing DA flows!  

A compare suite should be independent of anything - and it should be flexible for how the user organizes the data -or perhaps less generally, it should allow for the organization of the data as in SWELL, but also as in GEOS-JEDI. I would like for the compare to be something that can be deployed by the average user and the the monitoring group; both will have different purposes - ultimately, not initially.

Dooruk commented:
@mranst would it be possible to demonstrate this PR and how the comparisons are done at the end of our meeting today?

mranst commented:
> @mranst would it be possible to demonstrate this PR and how the comparisons are done at the end of our meeting today?

Yep, sounds good

Dooruk commented:
@mranst this is good for me. If you can add one or two examples on the usage under `Practical Examples` in the documentation that would be great and then we can merge it.

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst committed:
Fix for localensembleda

mranst committed:
Fix for localensembleda

mranst committed:
Use datetime interval

mranst committed:
Use datetime interval

mranst committed:
fixes for localensembleda

mranst committed:
fixes for localensembleda

mranst committed:
Merge branch 'feature/mranst/workflows_for_comparison' of discover.nccs.nasa.gov:/home/manstett/swell-main into feature/mranst/workflows_for_comparison

mranst committed:
Clean up eva comparison plots

mranst committed:
Clean up eva comparison plots

mranst committed:
Add Jedi-log comparison plots

mranst committed:
Add Jedi-log comparison plots

mranst committed:
fixes for jedi_log

mranst committed:
fixes for jedi_log

mranst committed:
add legend

mranst committed:
add legend

mranst committed:
Add jedi log parser for atmos

mranst committed:
Merge branch 'feature/mranst/workflows_for_comparison' of discover.nccs.nasa.gov:/home/manstett/swell-main into feature/mranst/workflows_for_comparison

mranst committed:
add increment for atmos

mranst committed:
fix

mranst committed:
fix transform

mranst committed:
Fix layout

mranst committed:
add examples

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst commented:
I cleaned up the comparison tasks and added one that creates jedi output plot comparisons for two experiments. I also added increment comparisons for geos_atmosphere. You can now just run `compare_variational_marine` and `compare_variational_atmosphere`, rather than `compare_3dvar`.

Dooruk commented:
i would like to focus on this next, this is some substantial update though. 

This also includes significant flow.cylc changes. Maybe we should have a meeting to discuss them, there is a limited amount of people involved in the flow.cylc modification.

mranst commented:
> i would like to focus on this next, this is some substantial update though.
> 
> This also includes significant flow.cylc changes. Maybe we should have a meeting to discuss them, there is a limited amount of people involved in the flow.cylc modification.

Sounds good. We're going to have a lot going on at the next swell meeting, it might make sense to schedule an additional meeting in the next week or so

mranst committed:
remove flow.cylc

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst committed:
add example to docs

mranst committed:
add eva_capabilities workflow

mranst committed:
correct eva_capabilities workflow

mranst committed:
fix sstagejedicycle

mranst committed:
replace defer_to_platform

mranst committed:
test fixes

mranst committed:
simplify comparison suite

mranst committed:
add tier test output

mranst committed:
clean up tier test output

mranst committed:
write to file

mranst committed:
add new documentation

mranst commented:
@Dooruk @ashiklom This should be ready for another review pass. I've simplified some of the comparison suite, and added another task called `TierTestOutput` which parses jedi log analysis and formats it into a comparison table, which looks like this:
 
<img width="1454" height="290" alt="image" src="https://github.com/user-attachments/assets/f8802705-cafc-400a-8b3d-794b0c7c782b" />

I've added some documentation for the flow.cylc changes under `docs/examples/templating_workflows.md`, and also some for the comparison workflow under `docs/examples/soca/comparison_workflows.md`

mranst committed:
Add docs for csh interactive install (#631)

* Initial commit

* Update csh install instructions

* Update installing_swell_csh_interactive.md

* Update installing_swell_csh_interactive.md

* fix order of steps

* Update installing_swell_uv_venv.md

Dooruk commented:
> @Dooruk @ashiklom This should be ready for another review pass. I've simplified some of the comparison suite, and added another task called `TierTestOutput` which parses jedi log analysis and formats it into a comparison table, which looks like this:
> <img alt="image" width="1454" height="290" src="https://private-user-images.githubusercontent.com/178604374/492913059-f8802705-cafc-400a-8b3d-794b0c7c782b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3MjExMzEsIm5iZiI6MTc1ODcyMDgzMSwicGF0aCI6Ii8xNzg2MDQzNzQvNDkyOTEzMDU5LWY4ODAyNzA1LWNhZmMtNDAwYS04YjNkLTc5NGIwYzdjNzgyYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyNFQxMzMzNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNjk3YzQ2NTUyMmFjZGQxZmRlZmMwNDM1MjlhZjdmYTllMmUzNWExYTczODdjZDRiZjIzZWVlZTJlMjYyZWY1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7ua6k7YQKREC2WJifgx_X5j249ZRt536WcOW1wkHZuk">
> 
> I've added some documentation for the flow.cylc changes under `docs/examples/templating_workflows.md`, and also some for the comparison workflow under `docs/examples/soca/comparison_workflows.md`

I like this output. Maybe we could add a third column for the difference? Then we could use that output for enforcing zero-diff (well, with a reasonable tolerance) for Tier 2 nightly tests (not necessarily for this PR)?

mranst committed:
Add diff and tolerances for residual norms

mranst committed:
clean up

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst committed:
test fixes

mranst commented:
> I like this output. Maybe we could add a third column for the difference? Then we could use that output for enforcing zero-diff (well, with a reasonable tolerance) for Tier 2 nightly tests (not necessarily for this PR)?

👍 Can set tolerances for different values

<img width="960" height="224" alt="image" src="https://github.com/user-attachments/assets/9efb5d29-c77b-4d3c-8dcc-e0eaf6a71fe0" />

mranst committed:
Merge branch 'develop' into feature/mranst/workflows_for_comparison

mranst committed:
add log comparison

Dooruk commented on docs/examples/templating_workflows.md:
```diff
@@ -0,0 +1,106 @@
+
+The `define_graph_section` task is used to set the graph. There are a few built-in methods used to format the strings into cylc syntax. The `format_cycle` method is used to construct properly indented blocks for cycling intervals. `create_new_section` creates a section object that tracks indentation levels, and can be added to other sections, with properly handled indentation and spacing. Here th ...
+
+## Tasks and the runtime section
```
Really appreciate the documentation, it helps me quite a bit to follow the changes.

One thing I'm wondering about how would we be able to add other types of potential entries for runtime tasks with this new approach?

Some examples for `RunGeosExecutable` below:

```cylc
        [[[job]]]
            shell = /bin/csh
```

```cylc
execution_time_limit
```

https://github.com/GEOS-ESM/swell/blob/f9fb2ae6946a82e494bbdbbff250eb2a2a2ae29c/src/swell/suites/3dvar_cycle/flow.cylc#L172

Dooruk commented on docs/examples/templating_workflows.md:
````diff
@@ -0,0 +1,106 @@
+                                           'nccs_discover_cascade': 2}})
+```
+
+On the `nccs_discover_cascade` platform, `nodes` will be set as 2, but on any other platform it will be 1. User overrides will still work as they did previously.
````
very cool feature!

Dooruk commented on src/swell/tasks/jedi_log_comparison.py:
```diff
@@ -0,0 +1,154 @@
+
+comparison_fields = {'Residual norm': {'delimiter': '=', 'dtype': float}}
+
+tolerances = {'Residual norm ( 0)': 0.01,
```
Number of the Residual norm array is dynamic and defined by the `number_of_iterations`

Dooruk reviewed (commented):
This is shaping up great. I think it is very close to merging, I would like others' input on this and I will bring it up during the development meeting.

I will make one majorish comment regarding the log file names. For compare variational, only `jedi_variational_log` is used here. For FGAT, which is another flavor of variational, log name would be `jedi_fgat_log`. One way to handle this might be to have another key in the override such as `comparison_log_type`, which would then dictate this in the EvaComparion tasks or `JediOopsLogParser` or some of the eva YAMLs.

then maybe we could define another comparison workflow within the same suite as such:

```py
    compare_fgat_marine = QuestionList(
        list_name="compare_variational_marine",
        questions=[
            compare_variational,
            qd.comparison_log_type('fgat'),
            qd.model_components(['geos_marine']),
        ]
    )
```

For `localensembleda` the log structure will be quite different so we don't have to worry about that for now, but my suggestion would also help with that going forward. @metdyn might shed some insight for OOPS log outputs for `localensembleda`.

Finally, I have to say, changes in `flow.cylc` are a bit confusing but I understand it will provide us a lot more flexibility in some regards and maybe to the extent we might be able to use another workflow management engine (?). Either way, only a handful of developers would be messing with that part so we can just explain to them, or expand upon the documentation as needed. 

Thanks again for working on this, this really provides substantial amount of changes.
